h/streamer/websocket.py

```diff
@@ -229,3 +235,56 @@ def _expand_uris(session, clause):
         expanded.update(storage.expand_uri(session, item))
 
     clause["value"] = list(expanded)
+
+
+class NumericMetric:
+    """Helper for tracking changes in a numeric metric over time."""
+
+    def __init__(self, get_metric):
+        """
+        :param get_metric: Callable which returns the current value of the metric
+        """
+        self.get_metric = get_metric
+        self.prev_value = get_metric()
+
+    def delta(self):
+        """Update the metric and return the difference between current and previous values"""
+        new_value = self.get_metric()
+        delta = new_value - self.prev_value
+        self.prev_value = new_value
+        return delta
+
+
+@newrelic.agent.data_source_factory(name="WebSocket Connections")
+def websocket_metrics(settings, environ):
+    """
+    A New Relic metric data source which provides metrics about WebSocket
+    connections.
+
+    See https://docs.newrelic.com/docs/agents/python-agent/supported-features/python-custom-metrics.
+    """
+
+    messages_sent_metric = NumericMetric(
+        lambda: sum([ws.messages_sent for ws in WebSocket.instances])
+    )
+    messages_received_metric = NumericMetric(
+        lambda: sum([ws.messages_received for ws in WebSocket.instances])
+    )
+    prefix = "Custom/WebSocket"
```

Comment:
The naming scheme for the metrics follows the examples in the [New Relic docs](https://docs.newrelic.com/docs/agents/python-agent/supported-features/python-custom-metrics) linked above. In particular, the `Custom/` prefix [is required](https://docs.newrelic.com/docs/agents/python-agent/supported-features/python-custom-metrics#naming-of-custom-metrics).